src/board/internal.tsx

```diff
@@ -112,6 +112,15 @@ export function InternalBoard<D>({
     );
   }
 
+  const itemsLayoutRef = useRef(itemsLayout);
+  itemsLayoutRef.current = itemsLayout;
+
+  // Updating layout if it changes during an active transition which can happen e.g. due to
+  // the appearing scrollbar when the transition starts.
+  useEffect(() => {
+    dispatch({ type: "update-layout", itemsLayout: itemsLayoutRef.current });
+  }, [columns]);
```

Comment:
This solves a symptom, but not the root cause.

Root cause – the layout engine uses stale `itemsLayout`. In this PR you fixed it for one-off case, but the architecture issue remains.

Can you think of any solution to the root cause issue?

Reply:
> I can't think of a good solution that wouldn't involve a significant rewrite of the existing code.

You will have to do it sooner or later, because this bugs will keep popping up. Instead of storing layout in the transition, you can always apply the latest layout.  

The choice is yours, if you want to keep the current way or sit down and refactor it properly once and for long time.

Reply:
Applying the latest layout is also challenging as it is unclear what the component should do exactly when the layout changes in the middle of a transition. The current solution is to ignore it (with an exception of when the columns change).

The ideal solution would be to listen to all such events and deliberately act on those, suppress or queue.

For now I would prefer proceeding with the fix as it fixes the real bug we are aware of.

Reply:
Layout shift calculation used to be very deterministic:

layout shift = widget original positions + current place to claim - original widget place (if not adding)

If the code does something else, the code needs to change, not the original design

Reply:
So, what exactly should happen when during the transition some items are added or removed, including the target one? Or if the layout changes from 4-col to 2-col in the middle of the operation?

Reply:
Thanks to our discussion I managed to come up with a cleaner solution: when a transition is present, use source layout from the transition. Otherwise - use current layout from props.items and observed columns. Wdyt?

Reply:
Answering this comment https://github.com/cloudscape-design/board-components/pull/244#discussion_r1233660721, because we need to be aligned on system design decisions.

Layout is a pure function, so it always uses the latest state of the props, even if it is weird (but valid according to our contract).

This may cause issues with scrollbars, but these are half-expected, in fact. Can give more info in the team chat for better knowledge sharing.

Reply:
We can make it work so that any change would transition somewhat except for maybe the target being missing. However, I see no point in doing so as the results would be rather random.

The current change synchronises the layout model with what is actually rendered on the screen. For example, if you start a transition and then resize the window while in transition, the layout will not resize at that point but will resize once the transition is over which is I think quite reasonable.

Reply:
Well, there is some intersection. The layout calculation itself is a pure function. However, to it takes the path property as input which is obtained based on the pointer moves over the placeholders. That path would make no sense if the layout changes. In that case, it is possible to fallback to the default path (the shortest) to the current pointer position but that would make little sense as the item positions between the layouts are different and the resulting path would not represent the user intent well.